File: js/plugins/middleware/examples/tool_approval.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { googleAI } from '@genkit-ai/google-genai';
import { genkit, z } from 'genkit';
import { toolApproval } from '../src/index.js';

const ai = genkit({
  plugins: [googleAI()],
});

const weatherTool = ai.defineTool(
  {
    name: 'weather',
    description: 'get the weather',
    inputSchema: z.object({ city: z.string() }),
    outputSchema: z.string(),
  },
  async ({ city }) => {
    return `The weather in ${city} is nice.`;
  }
);

async function main() {
  console.log('--- Running without approval ---');
  const response = await ai.generate({
    model: googleAI.model('gemini-3-flash-preview'),
    tools: [weatherTool],
    prompt: 'What is the weather in Paris?',
    use: [toolApproval()],
  });

  if (response.finishReason === 'interrupted') {
    console.log('Interrupted!');
    const interrupt = response.interrupts[0];
    console.log('Interrupt metadata:', interrupt?.metadata?.interrupt);
    if (!interrupt) {
      throw new Error('No interrupt found');
    }

    // approve the tool:
    interrupt.metadata = { ...interrupt.metadata, 'tool-approved': true };

    console.log('\n--- Running with approval ---');
    const response2 = await ai.generate({
      model: googleAI.model('gemini-3-flash-preview'),
      messages: response.messages,
      tools: [weatherTool],
      use: [toolApproval()],
      resume: {
        restart: [interrupt],
      },
    });
    console.log('Response:', response2.text);
  } else {
    console.log('Response:', response.text);
    return;
  }
}

main().catch(console.error);

File: py/packages/genkit-middleware/src/genkit_middleware/_tool_approval.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Tool approval middleware for Genkit."""

from __future__ import annotations

import json
from collections.abc import Awaitable, Callable

from pydantic import BaseModel, Field

from genkit._ai._tools import Interrupt
from genkit._core._action import ActionKind
from genkit._core._tracing import SpanMetadata, run_in_new_span
from genkit.middleware import BaseMiddleware, GenerateMiddlewareContext, MultipartToolResponse, ToolHookParams


class ToolApprovalConfig(BaseModel):
    """Tools that may run without an approval interrupt."""

    allowed_tools: list[str] = Field(default_factory=list)


class ToolApproval(BaseMiddleware[ToolApprovalConfig]):
    """Tool approval middleware that interrupts execution for non-allowed tools."""

    async def wrap_tool(
        self,
        params: ToolHookParams,
        ctx: GenerateMiddlewareContext,
        next_fn: Callable[[ToolHookParams, GenerateMiddlewareContext], Awaitable[MultipartToolResponse]],
    ) -> MultipartToolResponse:
        """Intercept tool execution and require approval if not in allowed list."""
        tool_name = params.tool.name

        if tool_name in self.config.allowed_tools:
            return await next_fn(params, ctx)

        metadata = params.tool_request_part.metadata or {}
        resumed = metadata.get('resumed')
        if isinstance(resumed, dict) and (resumed.get('toolApproved') or resumed.get('tool_approved')):
            return await next_fn(params, ctx)

        tool_input = params.tool_request_part.tool_request.input
        with run_in_new_span(
            SpanMetadata(name=tool_name, type='action', subtype=ActionKind.TOOL, input=tool_input),
        ) as span:
            if tool_input is not None:
                inp_json = tool_input.model_dump_json() if isinstance(tool_input, BaseModel) else json.dumps(tool_input)
                span.set_attribute('genkit:input', inp_json)
            raise Interrupt({'message': f'Tool not in approved list: {tool_name}'})
